rfcs/plugins.md

```diff
@@ -0,0 +1,349 @@
+   another: Joi.string().default('foo'),
+}
+```
+
+Will allow the following section in `config.toml`:
+
+```toml
+[plugin.my-plugin-name]
+someKey = true
+some_other_key = 'some kinda string'
+```
+
+The values from this config section will be passed to both the
+`processConfig` function for commands, as well as the
+[`activate`](#activate) function for server plugins. The object passed
+will look like:
+
+```js
+{ someKey: true,
+  some_other_key: 'some_kinda_string',
+  another: 'foo',
+}
+```
+
+### HTTP route
+
+The `httpRoute` plugin property is an object with a single key
+defining the http route the plugin will receive requests from. The
+value is the request handler for the route. The function receives
+three arguments:
+ - An Express `request` object
+ - An Express `response` object
+ - A state object of whatever kind is returned by the plugin's
+   [activate](#activate) function
+
+The handler should deal with whatever HTTP methods come to the route.
+
+Example:
+
+```js
+httpRoute: {
+    'hello-world': (req, res, state) => {
+        res.send(`Hello ${state.value}`);
+            res.end()
+    },
+}
+```
+
+Notes:
+ - The handler is responsible for calling `res.end()`.
+ - The route is a hard-prefix from the root of the horizon server address.
+  - It cannot clash with any other other loaded plugin routes (no
+    overlapping)
+      - If multiple routes need to be handled, the r
+      - Example, if you want to create a REST plugin, you need to
+        specify the route as something like `rest` and the endpoints
+        underneath it like `/rest/users` `/rest/vehicles` etc.
+  - It cannot clash with the route horizon uses itself (usually
+    `"horizon"`)
+ - In the configuration, the user may override the httpRoute for the
+   plugin, and the plugin must behave properly.
+ - Express 4.0 api for request and response objects
+
+### Processes
+
+The `processes` plugin property is an object with keys that correspond
+to process titles, and values which are objects with 2 properties:
+ - `run`: a function receiving the plugin state that runs as the main
+   thread of the process. It should not spawn processes itself
+ - `cleanup`: a function receiving the plugin state that cleans up the
+   process
+
+### Environment vars
+
+The `env` plugin property is an object from keys with environment
+variables, and Joi validators for values.
+
+Example:
+```js
+env: {
+  'MY_ENV_VALUE': Joi.any().allow('yes', 'no').default('no')
+}
+```
+
+The values of env variables that successfully validate are passed to
+the [activate](#activate) function as well as to the `processConfig` function for any [command](#command).
+
+**Note**:
+
+The env vars declared here will be validated to ensure no other plugin
+declares them. If a plugin wants to make use of externally defined env
+variables like `$PATH` etc, and doesn't want to mutual exclusivity to be enforced by Horizon, it can just access `process.env` like normal.
+
+### Validation helpers
```

Comment:
Technically, yes. It can run any Node code. So while we'd probably suggest that it be side-effect free, you can make something horribly inefficient. Of course the user opts into calling it, so they can opt out if they want